File: days/day01/solution.rb

def solve_first(inp)
  inp.lines.map { |num| num.to_i / 3 - 2 }.sum
end


def fuel_required(mod)
  remaining = mod / 3 - 2
  return 0 if remaining <= 0
  remaining + fuel_required(remaining)
end


def solve_second(inp)
  inp.lines.map { |num| fuel_required num.to_i }.sum
end


test = "\
14
1969
100756\
"
input = File.open('input.txt').read

puts solve_first input
puts solve_second input

File: days/day01/solution.hs
toInt :: String -> Integer
toInt s = read s :: Integer

required :: Integer -> Integer
required mass = mass `div` 3 - 2

compute :: (Integer -> Integer) -> String -> Integer
compute f inp = sum $ map (f . toInt) $ lines inp

first :: String -> Integer
first = compute required

fuelRequired :: Integer -> Integer
fuelRequired mass =
  if remaining <= 0
    then remaining
    else remaining + fuelRequired remaining
  where
    remaining = required mass

second :: String -> Integer
second = compute fuelRequired

main :: IO ()
main = do
  testInput <- readFile "test-input.txt"
  input <- readFile "input.txt"
  print $ first input
  print $ second input
